### PR TITLE
FEATURE: granular webhooks

### PR DESCRIPTION
**Pull request by lis2**
Before this change, webhooks could be only configured for specific groups like for example, all topic events. 

We would like to have more granular control like for example `topic_created` or `topic_destroyed`.

Test are failing because plugins changed has to be merged as well:
https://github.com/discourse/discourse-assign/pull/498
https://github.com/discourse/discourse-solved/pull/248
https://github.com/discourse/discourse-topic-voting/pull/159

Screenshot 2023-08-15 at 4 03 46 pm

### CODE REVIEW

File: app/models/web_hook_event_type.rb
Comment (martin-brennan) [inline app/models/web_hook_event_type.rb:4]: This scares me a little...even with the migration. I think maybe it would be safer to leave these in this PR, then make a followup PR to delete them.

File: app/models/web_hook_event_type.rb
Comment (martin-brennan) [inline app/models/web_hook_event_type.rb:5]: Nice that you have made these jump up in increments of 100 for each group :+1:

File: app/models/web_hook_event_type.rb
Comment (martin-brennan) [inline app/models/web_hook_event_type.rb:84]: All these things are bleh (not your fault) -- we should not be referring to plugins like this in core, but rather having a plugin API to register them. Maybe make a internal TODO for this?

File: config/locales/client.en.yml
Comment (martin-brennan) [inline config/locales/client.en.yml:4913]: IMO a nicer way of doing this to preserve the heirarchy would be to do this:

```
topic_event:
    group_name: "Topic Events"
    created: "Topic is created"
    ....
```

Comment (martin-brennan) [inline config/locales/client.en.yml:4913]: IMO a nicer way of doing this to preserve the heirarchy would be to do this:

```
topic_event:
    group_name: "Topic Events"
    created: "Topic is created"
    ....
```

File: app/models/web_hook_event_type.rb
Comment (martin-brennan) [reply app/models/web_hook_event_type.rb:4]: But of course still make your new `TYPES` const at the same time, I just mean leave these old ones in for now.

File: db/migrate/20230807033021_add_group_to_web_hook_event_type.rb
Comment (martin-brennan) [inline db/migrate/20230807033021_add_group_to_web_hook_event_type.rb:5]: Personally I think this should be an `:integer` and then have an Enum for the different groups in the model, then we are not repeating string data tons of times in the DB. Then you could have `WebHookEventType.groups[:topic]`

File: db/migrate/20230807040058_move_web_hooks_to_new_event_ids.rb
Comment (martin-brennan) [inline db/migrate/20230807040058_move_web_hooks_to_new_event_ids.rb:6]: Is the table name incorrect here? I don't have `web_hook_event_types_hooks` in my DB...

File: db/fixtures/007_web_hook_event_types.rb
Comment (martin-brennan) [inline db/fixtures/007_web_hook_event_types.rb:4]: IMO it's a little awkward to have to type out this kind of lookup all the time (`WebHookEventType::TYPES[:topic][:created]`), too many brackets and colons...will try to think of something else. Not a huge deal if it's just merged as is tho

File: app/controllers/admin/web_hooks_controller.rb
Comment (martin-brennan) [inline app/controllers/admin/web_hooks_controller.rb:21]: Could have a `scope` on `WebHookEventType` for this called `active_grouped`

**Review by martin-brennan — COMMENTED** [review]
I have left some comments but ran out of time before our call...will come back for a second pass

File: app/assets/javascripts/admin/addon/routes/admin-web-hooks-edit.js
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/routes/admin-web-hooks-edit.js:19]: Why is this necessary, can't you just rely on the boolean in the hbs?

File: db/fixtures/007_web_hook_event_types.rb
Comment (martin-brennan) [reply db/fixtures/007_web_hook_event_types.rb:4]: Maybe something like this in the model?


```ruby

  def self.define_type_methods
    TYPES.each do |type, events|
      events.each do |event_name, event_value|
        define_singleton_method("#{type}_#{event_name}") { TYPES[type][event_name] }
      end
    end
  end

  define_type_methods
```

Then you can do this:

```ruby
brc
Loading development environment (Rails 7.0.7)
[1] pry(main)> WebHookEventType.topic_created
=> 101
[2] pry(main)> WebHookEventType.user_promoted_created
=> 1601
[3] pry(main)>
```

File: db/migrate/20230807040058_move_web_hooks_to_new_event_ids.rb
Comment (martin-brennan) [reply db/migrate/20230807040058_move_web_hooks_to_new_event_ids.rb:6]: Ah I found it  https://github.com/discourse/discourse/blob/f17b70ede33e44da5a5adf723d4fb312ce92a3ea/db/migrate/20160905085445_create_join_table_web_hooks_web_hook_event_types.rb , we should have a model file for this, can you make one please so it's more obvious?

So I guess what we are saying now is:

* Previously there was 1 entry in this table, `TOPIC` which is `web_hook_event_type_id` of `1`, which represented a wildcard
* After this migration there will be `5` records in this table, one for each of the `TYPES[:topic]` events per web hook, since those 5 make up the wildcard now?

Maybe we should just do a migration spec too to confirm expected outcome.

File: spec/fabricators/web_hook_fabricator.rb
Comment (martin-brennan) [inline spec/fabricators/web_hook_fabricator.rb:11]: I think a pattern like this would be better since then we don't have to do 5 DB calls every time we fabricate:

```ruby
  after_build do |web_hook, transients|
    event_types =
      WebHookEventType.where(name: %w[post_created post_edited post_destroyed post_recovered])
    web_hook.web_hook_event_types = [
      transients[:post_created_hook] || event_types.find { |e| e.name == "post_created" },
      transients[:post_edited_hook] || event_types.find { |e| e.name == "post_edited" },
      transients[:post_destroyed_hook] || event_types.find { |e| e.name == "post_destroyed" },
      transients[:post_recovered_hook] || event_types.find { |e| e.name == "post_recovered" },
    ]
  end
```

File: spec/requests/admin/web_hooks_controller_spec.rb
Comment (martin-brennan) [inline spec/requests/admin/web_hooks_controller_spec.rb:23]: Should use the actual const for these from TYPES

File: app/models/web_hook_event_type.rb
Comment (lis2) [reply app/models/web_hook_event_type.rb:84]: Agree, I will create a TODO after this PR is merged (I don't want someone to pick it up before this change is finished)

File: db/fixtures/007_web_hook_event_types.rb
Comment (lis2) [reply db/fixtures/007_web_hook_event_types.rb:3]: We have overwritten db:migrate rake tasks and include run seed there: https://github.com/discourse/discourse/blob/main/lib/tasks/db.rake#L253

File: db/fixtures/007_web_hook_event_types.rb
Comment (lis2) [reply db/fixtures/007_web_hook_event_types.rb:4]: I am not that big fan of metaprogramming unless benefits are clearly visible. In that case, TYPES constant can be easily found and developer s clearly understand nested hash. Defining methods on the fly is much harder to grep. I read once about the idea of putting a comment about defined methods to make it searchable, like:
```
# def topic_created
# def user_promoted_created
```

File: db/migrate/20230807040058_move_web_hooks_to_new_event_ids.rb
Comment (lis2) [reply db/migrate/20230807040058_move_web_hooks_to_new_event_ids.rb:6]: Good idea about missing model file for many-to-many table 🚀 

Logic about number is a bit different.

Before `1` represented all topic events. And now we have 1 - topic created, 2 - topic updated etc. 

This migration is solving only individual events. 

Wildcard webhooks are untouched and they are represented by `WebHook.wildcard_web_hook` boolean table.

Logic to decide and webhook is active is here: https://github.com/discourse/discourse/blob/main/app/models/web_hook.rb#L46

File: app/models/web_hook_event_type.rb
Comment (martin-brennan) [reply app/models/web_hook_event_type.rb:132]: @lis2 I think you could do this now if you haven't yet, this PR should be resolved soon :)

File: db/fixtures/007_web_hook_event_types.rb
Comment (martin-brennan) [reply db/fixtures/007_web_hook_event_types.rb:4]: I hear your concern about this, and I do somewhat agree since this often annoys me too, we need to strike a balance between different devex concerns, and is resilient to typos.

Maybe we can just make the scope part of the type, instead of another level of the hash, so we change to this:

```ruby
# old
  b.id = WebHookEventType::TYPES[:chat][:message_restored]
# new
  b.id = WebHookEventType::TYPES[:chat_message_restored]
```

That way at least it's less awkward to type out than the nested way.

**Review by martin-brennan — APPROVED** [review]
Thanks for making those changes Kris, approving now though I do think we could improve the `WebhookEventType::TYPES[][]` devex a bit, please just take a look at my comment there. Also we are failing specs here for plugins, is there a companion PR?

**Comment by lis2** [comment]
@martin-brennan thank you for review :) I will change those nested TYPES. Corresponding PRs are:
https://github.com/discourse/discourse-assign/pull/498
https://github.com/discourse/discourse-solved/pull/248
https://github.com/discourse/discourse-topic-voting/pull/159

**Comment by martin-brennan** [comment]
@lis2 thanks I approved all the plugin PRs, **however** you will also need to make .discourse-compatibility entries for each one.

**Pull request merged**
